reStream.sh

```diff
@@ -86,6 +79,65 @@ if ! ssh_cmd true; then
     exit 1
 fi
 
+rm_version="$(ssh_cmd cat /sys/devices/soc0/machine)"
+
+# We cannot read the framebuffer directory in the reMarkable 2, but we know the
+# address of the pointer to the framebuffer for some reMarkable 2 versions.
+# This will need to be changed each time xochitl updates.
+rm2_getpointer() {
+    xochitl_version="$(ssh_cmd "sed -n 's/^REMARKABLE_RELEASE_VERSION=//p' /usr/share/remarkable/update.conf")"
+    case "$xochitl_version" in
+        "2.3.*")
+            echo "4121024"
+            ;;
+        "2.4.*")
+            # TODO
+            echo "???"
+            ;;
+        *) # use last known version if we cannot find a match
+            echo "WARNING: unknown reMarkable 2 release version" >&2
+            echo "consider updating if reStream doesn't work" >&2
+            echo "???"
+            ;;
+    esac
+}
+
+case "$rm_version" in
+    "reMarkable 1.0")
+        width=1408
+        height=1872
+        bytes_per_pixel=2
+        pixel_format="rgb565le"
+        # calculate how much bytes the window is
+        window_bytes="$((width * height * bytes_per_pixel))"
+        # read the first $window_bytes of the framebuffer
+        head_fb0="dd if=/dev/fb0 count=1 bs=$window_bytes 2>/dev/null"
+        ;;
+    "reMarkable 2.0")
+        width=1404
+        height=1872
+        bytes_per_pixel=1
+        pixel_format="gray8"
+        pid="$(ssh_cmd pidof xochitl)"
+        pointer="$(rm2_getpointer)"
+        read_address="addr=\$(dd if=/proc/$pid/mem bs=1 count=4 skip=$pointer 2>/dev/null | hexdump | awk '{print \$3\$2\}') && printf '%d' \$((16#\$addr))"
+        skipbytes="$(ssh_cmd "$read_address")"
+        # calculate how much bytes the window is
+        window_bytes="$((width * height * bytes_per_pixel))"
+        # carve the framebuffer out of the process memory
+        head_fb0="dd if=/proc/$pid/mem bs=1 count=$window_bytes skip=$skipbytes 2>/dev/null"
```

Comment:
There's a better solution to this that i'm trying to figure out.. Unfortantely i'm not great at bash. Basically you want to get the highest factor to 4096 that divides skip_bytes without a remainder. it'd be something like:

```bash
for((c=1;c<=4096;c++))
do
    if[ (($skip_bytes % c)) -eq 0 ] then
        div=$c
    fi
done
```
then `bs`=$div, `skip=(($skipbytes/$div))`, `count=(($window_bytes/$div))`

Reply:
believe you're looking for the closest integer to 4096 (the page block limit) that `$skipbytes` is divisible by... You're also going to _round-up_ count so that it captures more data than necessary (this gets filtered out by the 2nd `dd` command)

Reply:
Why don't you use `iflag=skip_bytes,count_bytes`? https://superuser.com/questions/380717/how-to-output-file-from-the-specified-offset-but-not-dd-bs-1-skip-n

Reply:
I think the fastest, though not pretty is simply to do:

```sh
num=1925402632

for i in {4096..1};
do
    if [ $(($num % $i)) == 0 ]; then
       echo "Largest divisor: $i"
       break
    fi
done
```
In my case, I'm getting 3284, which seems to work when updating @harrylepotter's gist.

Reply:
It seems I misunderstood the problem. Thomsten's solution seems efficient.

Reply:
@piada i think busybox's implementation of dd doesn't do `iflag`

Reply:
@harrylepotter @piada, yup, reading the implementation it doesn't seem like Busybox's implementation supports it: https://git.busybox.net/busybox/tree/coreutils/dd.c

Reply:
Thank you guys for your input. The problem with the gcd approach is that we will be way to dependent of the specific address for performance. I've added a `memread_instr` function to read the pages containing the buffer and then slice it to the desired bytes.